src/cogent3/core/new_alignment.py

```diff
@@ -3974,6 +4106,122 @@ def sliding_windows(
                 yield self[pos : pos + window]
 
+    def gapped_by_map(self, keep: FeatureMap, **kwargs):
+        # refactor: docstring
+        # todo: kath, not explicitly tested
+        seqs = {}
+        for seq in self.seqs:
+            selected = seq[keep]
+            seqs[seq.name] = numpy.array(selected.gapped_seq)
+
+        seqs_data = self._seqs_data.from_seqs(
+            data=seqs, alphabet=self.moltype.most_degen_alphabet()
+        )
+        return self.__class__(seqs_data=seqs_data, moltype=self.moltype, **kwargs)
+
+    def filtered(
+        self, predicate, motif_length: int = 1, drop_remainder: bool = True, **kwargs
+    ):
+        """The alignment positions where predicate(column) is true.
+
+        Parameters
+        ----------
+        predicate
+            a callback function that takes an tuple of motifs and returns
+            True/False
+        motif_length
+            length of the motifs the sequences should be split  into, eg. 3 for
+            filtering aligned codons.
+        drop_remainder
+            If length is not modulo motif_length, allow dropping the terminal
+            remaining columns
+        """
+        # refactor: type hint for predicate
+        # refactor: array
+        length = len(self)
+        if length % motif_length != 0 and not drop_remainder:
+            raise ValueError(
+                f"aligned length not divisible by motif_length={motif_length}"
+            )
+        gv = []
+        kept = False
+        seqs = [
+            self.get_gapped_seq(n).get_in_motif_size(motif_length, **kwargs)
+            for n in self.names
+        ]
+
+        positions = list(zip(*seqs))
+        for position, column in enumerate(positions):
+            keep = predicate(column)
+            if kept != keep:
+                gv.append(position * motif_length)
+                kept = keep
+
+        if kept:
+            gv.append(len(positions) * motif_length)
+
+        if not gv:
+            return None
+
+        locations = [(gv[i], gv[i + 1]) for i in range(0, len(gv), 2)]
+        # these are alignment coordinate locations
+        keep = FeatureMap.from_locations(locations=locations, parent_length=len(self))
+
+        return self.gapped_by_map(keep, info=self.info)
+
+    def omit_gap_pos(self, allowed_gap_frac: float = 1 - EPS, motif_length: int = 1):
+        """Returns new alignment where all cols (motifs) have <= allowed_gap_frac gaps.
+
+        Parameters
+        ----------
+        allowed_gap_frac
+            specifies proportion of gaps is allowed in each column. Set to 0 to
+            exclude columns with any gaps, 1 to include all columns.
+        motif_length
+            set's the "column" width, e.g. setting to 3 corresponds to codons.
+            A motif that includes a gap at any position is included in the
+            counting.
+        """
+
+        gaps = list(self.moltype.gaps)
+
+        gaps_ok = GapsOk(
```

Comment:
this GapsOk class may not be required for positions. This algorithm will be much faster with the proposed `iter_array_pos()` method. Add comment # redesign: use array methods